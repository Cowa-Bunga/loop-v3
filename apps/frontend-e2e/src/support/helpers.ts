export const testIdSelector = (testId: string) => `[data-testid="${testId}"]`

export const COOKIE_NAME = 'cowabunga-user-cookie'
