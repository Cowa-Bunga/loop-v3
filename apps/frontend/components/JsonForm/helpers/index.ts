const makeInputId = (id: string, suffix?: string) => {
  const parts = id.split('/')

  if (suffix) {
    return `${parts[parts.length - 1]}-${suffix}`
  }

  return parts[parts.length - 1]
}

export { makeInputId }
