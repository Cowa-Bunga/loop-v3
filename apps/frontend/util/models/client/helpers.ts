import { defaultsDeep } from 'lodash'
import { MODEL_CLIENT_EMPTY } from './model'

export const modelClientList = (
  data: { [s: string]: unknown } | ArrayLike<unknown>
): IClient[] => {
  return Object.entries(data)
    .map(modelClientData)
    .map((el: never) => el['1'])
    .sort((a: IClient, b: IClient) => a.name.localeCompare(b.name))
}

export const modelClientData = (data: unknown): IClient => {
  return defaultsDeep(data, MODEL_CLIENT_EMPTY)
}
