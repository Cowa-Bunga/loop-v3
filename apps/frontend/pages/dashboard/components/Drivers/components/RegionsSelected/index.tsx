import { useUserContext } from '@context/user'
import { collection, query, where } from '@util/lib/firebase'
import ui from '../../../Drivers/style'
import { Actions } from '../../../../actions'
import { stringAvatar } from '@util/helpers'
import {
  Alert,
  Avatar,
  Box,
  CircularProgress,
  Input,
  List,
  ListItem,
  ListItemText,
  Stack
} from '@mui/material'
import {
  useEffect,
  useFirestore,
  useFirestoreCollectionData,
  useMergeState
} from '@hooks'

export default function RegionSelected({ hubs }: { hubs: string[] }) {
  const { state: userContext } = useUserContext()

  const [state, setState] = useMergeState<{
    availableDrivers: IDriver[]
    busyDrivers: IDriver[]
  }>({
    availableDrivers: [],
    busyDrivers: []
  })

  const { modelDrivers } = Actions(state, setState)
  const driversRef = collection(useFirestore(), 'drivers')

  const contains = hubs.map((id) => ({
    promisor_id: userContext.client.client_id,
    hub_id: id
  }))

  const constraints = where(
    'delivery_permissions',
    'array-contains-any',
    contains
  )

  const q = query(driversRef, constraints)
  const { status, data } = useFirestoreCollectionData(q)

  useEffect(() => {
    if (status == 'success') {
      const drivers = modelDrivers(data)
      setState({
        availableDrivers: drivers.filter((d: IDriver) => d.available),
        busyDrivers: drivers.filter((d: IDriver) => !d.available)
      })
    }
  }, [data, status])

  return status !== 'success' ? (
    <Box sx={ui.container}>
      <CircularProgress color="secondary" />
    </Box>
  ) : (
    <Box sx={ui.container}>
      <Input
        sx={ui.searchDriver}
        fullWidth
        placeholder="Search driver name or code"
      />
      <Stack sx={ui.stack} spacing={2}>
        <Alert severity="success">Available</Alert>
        <List dense>
          {state.availableDrivers.map((driver) => (
            <ListItem key={driver.id}>
              <Avatar {...stringAvatar(driver.name)} />
              <div>&nbsp;</div>
              <ListItemText primary={driver.name} />
            </ListItem>
          ))}
        </List>
        <Alert severity="warning">Busy</Alert>
        <List dense>
          {state.busyDrivers.map((driver) => (
            <ListItem key={driver.id}>
              <Avatar {...stringAvatar(driver.name)} />
              <div>&nbsp;</div>
              <ListItemText primary={driver.name} />
            </ListItem>
          ))}
        </List>
      </Stack>
    </Box>
  )
}
