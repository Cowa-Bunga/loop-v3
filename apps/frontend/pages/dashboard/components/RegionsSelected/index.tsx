import { useUserContext } from '@context/user'
import { collection, query, where } from 'firebase/firestore'
import { useFirestore, useFirestoreCollectionData } from 'reactfire'
import ui from '../Drivers/style'
import { useMergeState, useEffect } from '@hooks'
import Actions from '../Drivers/actions'
import { stringAvatar } from '@util/helpers'
import {
  Alert,
  Avatar,
  Box,
  CircularProgress,
  Input,
  List,
  ListItem,
  ListItemText,
  Stack
} from '@mui/material'

export default function RegionSelected({ hubs }: { hubs: string[] }) {
  const { state: userContext } = useUserContext()

  const [state, setState] = useMergeState<{
    availableDrivers: IDriver[]
    busyDrivers: IDriver[]
  }>({
    availableDrivers: [],
    busyDrivers: []
  })
  const { modelDrivers } = Actions(state, setState)

  const driversRef = collection(useFirestore(), 'drivers')
  const contains = hubs.map((id) => ({
    promisor_id: userContext.client.client_id,
    hub_id: id
  }))
  const constraints = where(
    'delivery_permissions',
    'array-contains-any',
    contains
  )
  const q = query(driversRef, constraints)
  const { status, data } = useFirestoreCollectionData(q)

  useEffect(() => {
    if (status == 'success') {
      const drivers = modelDrivers(data)
      setState({
        availableDrivers: drivers.filter((d: IDriver) => d.available),
        busyDrivers: drivers.filter((d: IDriver) => !d.available)
      })
    }
  }, [data, modelDrivers, setState, status])

  return status !== 'success' ? (
    <Box sx={ui.container}>
      <CircularProgress color="secondary" />
    </Box>
  ) : (
    <Box sx={ui.container}>
      <Input
        sx={ui.searchDriver}
        fullWidth
        placeholder="Search driver name or code"
      />
      <Stack sx={ui.stack} spacing={2}>
        <Alert severity="success">Available</Alert>
        <List dense>
          {state.availableDrivers.map((driver) => (
            <ListItem key={driver.id}>
              <Avatar {...stringAvatar(driver.name)} />
              <div>&nbsp;</div>
              <ListItemText primary={driver.name} />
            </ListItem>
          ))}
        </List>
        <Alert severity="warning">Busy</Alert>
        <List dense>
          {state.busyDrivers.map((driver) => (
            <ListItem key={driver.id}>
              <Avatar {...stringAvatar(driver.name)} />
              <div>&nbsp;</div>
              <ListItemText primary={driver.name} />
            </ListItem>
          ))}
        </List>
      </Stack>
    </Box>
  )
}
