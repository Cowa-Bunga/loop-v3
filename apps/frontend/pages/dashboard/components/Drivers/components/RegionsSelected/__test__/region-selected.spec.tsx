import RegionsSelected from '../'
import { renderWithProviders } from '@test/helpers'

import { useFirestoreCollectionData } from '@hooks'

const mockUseFirestoreCollectionData = useFirestoreCollectionData as jest.Mock
jest.mock('@context/user', () => ({
  useUserContext: () => ({
    state: {
      client: {
        client_id: 'clientId'
      }
    }
  })
}))

jest.mock('@hooks', () => ({
  ...jest.requireActual('@hooks'),
  useFirestore: jest.fn(),
  useFirestoreCollectionData: jest.fn()
}))

jest.mock('@util/lib/firebase', () => ({
  collection: jest.fn(),
  query: jest.fn(),
  where: jest.fn(),
  firebaseConfig: {}
}))

describe('RegionsSelected', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })
  it('should render: loading', () => {
    mockUseFirestoreCollectionData.mockReturnValue({
      status: 'loading'
    })
    const hubs = ['hubId1', 'hubId2']
    const { baseElement } = renderWithProviders(<RegionsSelected hubs={hubs} />)
    expect(baseElement).toMatchSnapshot()
  })

  it('should render: success', () => {
    mockUseFirestoreCollectionData.mockReturnValue({
      status: 'success',
      data: [
        {
          name: 'test',
          email: 'test',
          mobile_no: 'test',
          vehicle_type: 'test',
          employee_code: 'test',
          location: 'test',
          lunch: 'test',
          available: 'test',
          on_active_trip: 'test',
          idle_since: 'test',
          created_at: 'test'
        }
      ]
    })
    const hubs = ['hubId1', 'hubId2']
    const { baseElement } = renderWithProviders(<RegionsSelected hubs={hubs} />)
    expect(baseElement).toMatchSnapshot()
  })
})
