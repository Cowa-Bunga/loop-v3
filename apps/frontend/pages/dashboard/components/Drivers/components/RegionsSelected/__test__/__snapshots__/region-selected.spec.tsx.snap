// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegionsSelected should render: loading 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-1yy4zhf"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorSecondary css-1szo49z-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>
</body>
`;

exports[`RegionsSelected should render: success 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-1yy4zhf"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth css-1ulwgiq-MuiInputBase-root-MuiInput-root"
      >
        <input
          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          placeholder="Search driver name or code"
          type="text"
          value=""
        />
      </div>
      <div
        class="MuiStack-root css-w9uisd-MuiStack-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardSuccess MuiAlert-standard css-wnkddu-MuiPaper-root-MuiAlert-root"
          role="alert"
        >
          <div
            class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="SuccessOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4C12.76,4 13.5,4.11 14.2, 4.31L15.77,2.74C14.61,2.26 13.34,2 12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0, 0 22,12M7.91,10.08L6.5,11.5L11,16L21,6L19.59,4.58L11,13.17L7.91,10.08Z"
              />
            </svg>
          </div>
          <div
            class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
          >
            Available
          </div>
        </div>
        <ul
          class="MuiList-root MuiList-padding MuiList-dense css-h4y409-MuiList-root"
        >
          <li
            class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding css-6lnq02-MuiListItem-root"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-9zf53r-MuiAvatar-root"
            >
              t
            </div>
            <div>
              Â 
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
              >
                test
              </span>
            </div>
          </li>
        </ul>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardWarning MuiAlert-standard css-to2f46-MuiPaper-root-MuiAlert-root"
          role="alert"
        >
          <div
            class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="ReportProblemOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
              />
            </svg>
          </div>
          <div
            class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
          >
            Busy
          </div>
        </div>
        <ul
          class="MuiList-root MuiList-padding MuiList-dense css-h4y409-MuiList-root"
        />
      </div>
    </div>
  </div>
</body>
`;
